### PR TITLE
replace boost filesystem exists

### PR DESCRIPTION
**Pull request by cybaol**
Task 3 of #5881

### CODE REVIEW

**Comment by mvieth** [comment]
Hi, my idea was not to replace _every_ `boost::filesystem::exists` with `std::ifstream(filename).good()`, but _only_ where a `std::ifstream` is opened for reading anyways. For example in io/src/ifs_io.cpp: after the check with `empty` and `exists`, the file is opened anyways with a `std::ifstream`, so we could remove `boost::filesystem::exists` there and add `!fs.good()` after opening.
In other cases, where the file is not opened as a `std::ifstream` immediately afterwards, I would not try to replace `boost::filesystem::exists` with `std::ifstream(filename).good()`. Opening a filestream just to check whether a file exists seems kind of wasteful, and I would assume that Boost's exists is more elegant.
I have come to the conclusion that it does not seem possible to replace _everything_ from boost::filesystem in PCL without using std::filesystem (C++17), at least not in an easy and elegant way. So I think it would be best to focus more on [idea 3 from the issue](https://github.com/PointCloudLibrary/pcl/issues/5881#issue-2006471333) (using both boost::filesystem and std::filesystem, and selecting the appropriate one with the preprocessor).

**Comment by cybaol** [comment]
The error seems a random failure.

**Comment by mvieth** [comment]
I believe it would make sense to first check whether the `file_name` is empty, before opening the filestream. That way we have made sure that the string contains something. We could also change the error message, for example "No file name given!" instead of "Could not find file ..." (or similar)

File: io/src/obj_io.cpp [hunk 1]
```diff
@@ -146,16 +146,16 @@ int
...
-  if (obj_file_name.empty() || !boost::filesystem::exists (obj_file_name))
+  if (obj_file_name.empty ())
   {
-    PCL_ERROR ("[pcl::MTLReader::read] Could not find file '%s'!\n",
-               obj_file_name.c_str ());
+    PCL_ERROR ("[pcl::MTLReader::read] No OBJ file name given!\n");
     return (-1);
   }
 
-  if (mtl_file_name.empty())
```
Comment (cybaol) [inline io/src/obj_io.cpp:156]: Move to `pcl::MTLReader::read (const std::string& mtl_file_path)`

**Comment by cybaol** [comment]
Random failure.

File: io/src/ifs_io.cpp [hunk 2]
```diff
@@ -60,15 +59,24 @@ pcl::IFSReader::readHeader (const std::string &file_name, pcl::PCLPointCloud2 &c
...
   std::uint32_t nr_points = 0;
+
+  if (file_name.empty ())
+  {
+    PCL_ERROR ("[pcl::IFSReader::readHeader] No file name given!\n");
+    return (-1);
+  }
+
+  // Open file in binary mode to avoid problem of
+  // std::getline() corrupting the result of ifstream::tellg()
```
Comment (mvieth) [inline io/src/ifs_io.cpp:70]: This comment does not make sense to me: `std::getline` is not used in ifs_io.cpp?

File: io/src/obj_io.cpp (hunk 1 above)
Comment (mvieth) [reply io/src/obj_io.cpp:156]: But removing this check here would change the behaviour: if `mtl_file_name` is empty here, then `mtl_file_path` below would contain `obj_file_path.parent_path()` (and maybe a slash or backslash at the end), meaning `mtl_file_path` would not be empty. So as I see it, it is necessary to keep this check here.

**Review by mvieth — APPROVED** [review]
Thanks!

**Pull request merged**